packages/core/src/RenderingEngine/StackViewport.ts

```diff
@@ -1708,14 +1710,25 @@ class StackViewport extends Viewport implements IStackViewport {
       }
 
       function sendRequest(imageId, imageIdIndex, options) {
-        return loadAndCacheImage(imageId, options).then(
-          (image) => {
+        const uncompressedIterator = ProgressiveIterator.as(
+          loadAndCacheImage(imageId, options)
+        );
+        const displayedIterator = new ProgressiveIterator<void | IImage>(
+          'displayed'
+        );
+        displayedIterator.generate(async (it) => {
+          for await (const image of uncompressedIterator) {
+            // Update cache
+            cache.removeImageLoadObject(imageId);
+            cache.putImageLoadObject(imageId, {
+              promise: Promise.resolve(image),
+            });
+            console.log('Adding new image to render', imageIdIndex, image);
             successCallback.call(this, image, imageIdIndex, imageId);
-          },
-          (error) => {
-            errorCallback.call(this, error, imageIdIndex, imageId);
+            it.add(image, uncompressedIterator.done);
           }
-        );
+        });
+        return displayedIterator.getDonePromise();
       }
```

Comment:
I'm uncertain about the current situation, but its significance may not be relevant at the moment. What truly matters is that I had no anticipation of encountering any progressive mention in either the StackViewport or the BaseVolumeViewport. However, I've noticed the presence of iterators and numerous elements such as decodeLevel, stageId, and others that strongly associate with HTJ2K, thereby compromising the independence of the viewport in relation to the loader.

All of these should be extracted and handled through event publishing and event handlers. Store your iterators in the dicomImageLoader, which is the appropriate entity to manage them, rather than the packages/core. Trigger events on the eventTarget and have event handlers listen to those events for cache removal or updates.